### DISCUSSION TITLE
chat-cli chat persistence

### DISCUSSION THREAD

**Discussion by WPrintz** [chat-cli/chat-cli#43] [Ideas]
Would it be possible to invoke a command within chat mode that would allow continuation of the chat from where you left off between different chat sessions?    For example, within chat mode could we command chat-cli like `> /save` and chat-cli would write-out an appropriate filetype that could be read into a new chat-cli chat session to continue from where the previous session ended?  For example, it could write a file named "chatcli-chat-2024-07-08T20:20:24Z.txt" 

If this file could be read into chat-cli chat mode when invoking the application, chat-cli could continue where the previous chat ended.

**Comment by micahwalter** [comment]
This is possible with a few caveats..

The main issue being that eventually the context window will get too long depending on the model you are using and will throw an error. It can also get costly with not a ton of benefit.

However, it would be nice to save previous chats.

We could do this via the logs mechanism built into Bedrock, or some alternative data store like a local SQLite db or just plain text files saved to disk. 

I have an open issue to utilize the logs. The main trick with this is the user has to enable them in Bedrock before they will work.

**Comment by micahwalter** [comment]
This is coming in the next release

**Comment by micahwalter** [comment]
OK, this is basically done, and I will merge it in with the next release later today. Here's the highlights:

1. Creates a config file on your local system (os dependent, but uses standard storage location for config)
2. Creates a local sqlite3 database on your system in the same folder as above - no need to install sqlite3
3. Chats created with the chat command are automatically saved in the DB as you create them. no need to type `save`
4. Use the `chat list` command to list most recent 10 chats (for now) - this will display the chat-id and first part of the original prompt
5. Use the `--chat-id` flag to load the chat like `chat-cli chat --chat-id 07927821-f443-4e92-84c6-86d6fa30ebf2`
6. Loading a chat will print out the loaded chat and allow you to pick up where you left off, and will continue saving to the same chat-id going forward
